### PR TITLE
Optimization

### PR DESCRIPTION
**Pull request by MichaelHeiss**
so I think the optimization file is working and I tried to document everything. would be nice to get some feedback on that.

### CODE REVIEW

**Comment by MichaelHeiss** [comment]
sorry i found some errors and had to fix those before the revue but i don't know whether i did this right. please excuse me.

File: bulletroboy/optimization.py [hunk 1]
```diff
@@ -0,0 +1,349 @@
...
+
+    ws_array_feasible, _ = workspace_feasible(radius=radius, height=height, angles=angles, forces=forces)
+    ws_array_feasible = (ws_array_feasible + (ws_array_feasible * WS_ARRAY_DESIRED * 9)) / 10
+    loss = np.dot((WS_ARRAY_DESIRED - ws_array_feasible), (WS_ARRAY_DESIRED - ws_array_feasible))
+    # print(f"Time: {time() - start} s")
+    print(loss)
+    print(param)
+    return loss
+
+#__main__
```
Comment (mlanas) [inline bulletroboy/optimization.py]: Would better to have an if main statement like:
`if __name__ == "__main__":`

File: bulletroboy/optimization.py [hunk 2]
```diff
@@ -0,0 +1,349 @@
+import numpy as np
+from os import path
+from scipy.optimize import minimize
+from utils import CageConfiguration
+from time import time
+import matplotlib.pyplot as plt
+# This import registers the 3D projection, but is otherwise unused.
+from mpl_toolkits.mplot3d import Axes3D  # noqa: F401 unused import
```
Comment (mlanas) [inline bulletroboy/optimization.py]: If the import is not used, it should be deleted

**Review by mlanas — CHANGES_REQUESTED** [review]

File: bulletroboy/optimization.py [hunk 3]
```diff
@@ -0,0 +1,349 @@
...
+    attachment_points[0, 4] = radius[1]
+    attachment_points[0, 5] = radius[1] * np.cos(angles[2])
+    attachment_points[0, 6] = radius[1] * np.cos(angles[3])
+    attachment_points[0, 7] = 0                             # attached on the person
+    attachment_points[2, 0:3] = height                      # upper tendons
+    attachment_points[2, 3] = 1.5                           # shoulder tendon
+    attachment_points[2, 4:7] = 0                           # lower tendons
+    attachment_points[2, 7] = 0.75                          # hip tendon
+
+    if joint == "human/left_wrist":
```
Comment (missxa) [inline bulletroboy/optimization.py]: remove hardcoded names of the joints

File: bulletroboy/optimization.py [hunk 4]
```diff
@@ -0,0 +1,349 @@
...
+    x, y, z = np.meshgrid(np.arange(-step_size, 1 + step_size, step_size),
+                          np.arange(-1, 1 + step_size, step_size),
+                          np.arange( 0, 2 + step_size, step_size))
+
+    xx = x.reshape(-1)
+    yy = y.reshape(-1)
+    zz = z.reshape(-1)
+    ws_array_desired = np.array([])
+    ws = np.array([xx, yy, zz])
+    for i, (mx, my, mz) in enumerate(zip(xx, yy, zz)):
```
Comment (missxa) [inline bulletroboy/optimization.py]: add comments in this function explaining what you're doing and the magic numbers

File: bulletroboy/optimization.py [hunk 5]
```diff
@@ -0,0 +1,349 @@
...
+ax.legend()
+ax.set_xlabel('X axis')
+ax.set_ylabel('Y axis')
+ax.set_zlabel('Z axis')
+plt.title(f"Workspace {solver}")
+n = 1
+data = {n: [data_L, data_R]}
+
+# Save test data to .txt file
+text_file = open(f"../../Desktop/{solver}_02.txt", "wt")
```
Comment (missxa) [inline bulletroboy/optimization.py]: remove hardcoded path

File: bulletroboy/optimization.py [hunk 6]
```diff
@@ -0,0 +1,349 @@
...
+    # print(f"Time: {time() - start} s")
+    return ws_array_desired, ws_desired
+
+def workspace_feasible(step_size=0.2, joint="human/left_wrist", radius=np.array([1., 1.]), height=2.,
+                       angles=np.array([np.pi/3, 2*np.pi/3]), forces=np.array([[100., 0., 0.]])):
+    """ This function calculates the feasible workspace of the arm of a standing person
+
+    Args:
+        step_size (float):                  The size of the iteration steps
+        joint (str):                        Information about the left or right wrist
```
Comment (missxa) [inline bulletroboy/optimization.py]: passing joint names as strings is prone to errors or typos. it is not a good practice. if you need to distinguish between left and right you can pass a bool `isRight`.  in case you need link names, they should not be hardcoded in the optimizer, but read from a config/ yaml file or requested from a robot or human. worst case defined in the main function

File: bulletroboy/optimization.py [hunk 7]
```diff
@@ -0,0 +1,349 @@
...
+    #print(f"Time: {time() - start} s")
+    return forces
+
+def get_attachment_points(radius, height, angles, joint):
+    """ This function calculates the attachment points for the tendons.
+
+    Args:
+        radius (numpy.array):               The radius of the upper part of the cage and the lower part of the cage
+        height (numpy.array):               The height of the cage
+        angles (numpy.array):               Angles of the tendons with respect to the x-axis
```
Comment (missxa) [inline bulletroboy/optimization.py]: you mean between the angle and x-axis?

File: bulletroboy/optimization.py [hunk 8]
```diff
@@ -0,0 +1,349 @@
+import numpy as np
+from os import path
+from scipy.optimize import minimize
+from utils import CageConfiguration
+from time import time
+import matplotlib.pyplot as plt
+# This import registers the 3D projection, but is otherwise unused.
+from mpl_toolkits.mplot3d import Axes3D  # noqa: F401 unused import
+
+def force_decomposition(force_vector, force_point, tendons, attachment_points):
```
Comment (missxa) [inline bulletroboy/optimization.py]: why is `force_decomposition` defined in the optimization? it doesn't belong here and must be identical for every module where it's used

File: bulletroboy/optimization.py [hunk 9]
```diff
@@ -0,0 +1,349 @@
...
+
+        if solution.success:
+            forces = {id: force for id, force in zip(tendons, solution.x) if force > MIN_FORCE}
+    # else:
+    #     # print(f"WARNING: Cannot decompose the forces: [ {solution.message} ]")
+
+    #print(f"Time: {time() - start} s")
+    return forces
+
+def get_attachment_points(radius, height, angles, joint):
```
Comment (missxa) [inline bulletroboy/optimization.py]: why do you need to calculate attachment points of the tendons if the config file provides this information?

File: bulletroboy/optimization.py [hunk 10]
```diff
@@ -0,0 +1,349 @@
...
+        attachment_points[1, 2] = -radius[0] * np.sin(angles[3])
+        attachment_points[1, 3] = 0
+        attachment_points[1, 4] = 0
+        attachment_points[1, 5] = -radius[1] * np.sin(angles[2])
+        attachment_points[1, 6] = -radius[1] * np.sin(angles[3])
+        attachment_points[1, 7] = 0
+    return attachment_points
+
+def workspace_desired(step_size=0.2, joint="human/left_wrist"):
+    """ This function calculates the ergonomic workspace of the arm of a standing person
```
Comment (missxa) [inline bulletroboy/optimization.py:128]: calculates based on what?

File: bulletroboy/optimization.py [hunk 11]
```diff
@@ -0,0 +1,349 @@
...
+    ii = np.where(ws_array_feasible == 1)[0]
+    ws_feasible = ws[:, ii]
+    # print(f"Time: {time() - start} s")
+    return ws_array_feasible, ws_feasible
+
+def cost_function(param):
+    """ This function calculates the L-2 distance loss between a desired and a feasible workspace
+
+    Args:
+        param (numpy.array):                Vector containing position parameters for the tendon attachment points
```
Comment (missxa) [inline bulletroboy/optimization.py]: explain the order of the values in param

File: bulletroboy/optimization.py [hunk 12]
```diff
@@ -0,0 +1,349 @@
...
+# solver = 'COBYLA'
+# solver = 'SLSQP'
+# solver = 'trust-constr'
+# xatol (the absolut error in xopt between iterations that is acceptable for convergence is set
+# to 0.001 which is equivalent to 1mm
+solution = minimize(cost_function, initial_forces, method='Nelder-Mead',
+                    options={'maxiter': 100, 'initial_simplex': initial_simplex, 'xatol': 0.001})
+
+# Get the position parameters that minimize the cost function.
+print(solution)
```
Comment (missxa) [inline bulletroboy/optimization.py]: this is a general comment on all of your prints: please add a debug message to all the variables that you print. you know that it's solution printed here, but nobody else

**Review by missxa — CHANGES_REQUESTED** [review]

**Comment by missxa** [comment]
@MichaelHeiss please resolve conflicts with the base branch
